docs/architecture/adr-040-storage-and-smt-state-commitments.md

```diff
@@ -48,7 +48,7 @@ For data access we propose 2 additional KV buckets (namespaces for the key-value
 2. B2: `hash(key, value) → key`: a reverse index to get a key from an SMT path. Recall that SMT will store `(k, v)` as `(hash(k), hash(key, value))`. So, we can get an object value by composing `SMT_path → B2 → B1`.
 3. we could use more buckets to optimize the app usage if needed.
 
-Above, we propose to use a KV DB. However, for the state machine, we could use an RDBMS, which we discuss below.
+We propose to use a KV database for both `SS` and `SC` - each run by its own database instance. This design allows for the separation of `SS` and `SC` into different hardware units, providing support for more complex setup scenarios and improving DB performance (essentially this will create two store shards: one for `SS` and another for `SC`). Moreover, we will be able to configure databases for `SS` and `SC` separately.
```

Comment:
Could this behavior be configurable? i.e. it could either be one database or two databases? It seems like two db's defeats the goal of atomic commits. In a configurable scenario, someone could simply pass in two prefix stores of the same db to use one atomic db.

Reply:
That's how I've implemented it - see https://github.com/cosmos/cosmos-sdk/pull/9892

Reply:
I agree. Is there any known performance problem you're trying to address?

Reply:
improved performance would require multi hardware setup.

Reply:
Wait isn't using a separate database, one merkelized, one not, for state been shown to be a huge efficiency improvement? Like, I thought it was the main change clients like turbogeth do? <img width="973" alt="Screenshot 2021-09-17 at 9 56 27 AM" src="https://user-images.githubusercontent.com/6440154/133804484-7c5b1a09-8358-47ea-8f03-a39f8d288dac.png"> (Granted turbogeth's entire db

Basically all your state reads during execution don't have to do trie traversals along the merkle trie, which is guaranteed to not be cache efficient due to hashes being on every tree node. Whereas you can instead do your writes in a flat-map, and just at the end update the merkle paths in the merkelized store. You can re-use the same leaf storage for no additional data

Reply:
@ValarDragon  - the question is about physical DB. We are still going to use 2 different stores: `SS` and `SC` (which is the main source of optimization).

Reply:
@aaronc if we will split DB into 2 hardware units, then we will have another performance boost. However, you are right - it will not make the atomic updates.

Reply:
During the call we are deciding to design the software to work with 2 DB instances as well as a single one.